File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/bstFromPreorder_1008.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class bstFromPreorder_1008 {
    int i = 0;

    public TreeNode bstFromPreorder(int[] preorder) {
        return helper(preorder, Integer.MAX_VALUE);
    }

    // bound 当前节点的父节点的值 【BST的左子树 < 父节点的值  右子树 > 父节点的值】
    private TreeNode helper(int[] preorder, int bound) {
        if (i == preorder.length || preorder[i] > bound) return null;
        TreeNode root = new TreeNode(preorder[i++]);
        root.left = helper(preorder, root.val);
        root.right = helper(preorder, bound);
        return root;
    }
}

File: Clion/ZeroTrac/1500_1600/bstFromPreorder_1008.cpp
//
// Created by TzashiNorpu on 7/18/2022.
//
#include "header.h"

using namespace ZeroTrac;

TreeNode *build(vector<int> &A, int &i, int bound) {
    if (i == A.size() || A[i] > bound) return NULL;
    TreeNode *root = new TreeNode(A[i++]);
    root->left = build(A, i, root->val);
    root->right = build(A, i, bound);
    return root;
}

TreeNode *bstFromPreorder_1008::bstFromPreorder(vector<int> &preorder) {
    int i = 0;
    return build(preorder, i, INT_MAX);
}


